gensim/models/nmf_pgd.pyx

```diff
@@ -0,0 +1,224 @@
+        double[::1] r_actual_data,
+        double lambda_,
+        double v_max
+    ):
+    """Bound new residuals.
+
+    Parameters
+    ----------
+    r_indptr : vector
+    r_indices : vector
+    r_data : vector
+    r_actual_indptr : vector
+    r_actual_indices : vector
+    r_actual_data : vector
+    lambda_ : double
+    v_max : double
+
+    Returns
+    -------
+    float
+        Cumulative difference between previous and current residuals vectors.
+    """
+
+    cdef Py_ssize_t n_samples = r_actual_indptr.shape[0] - 1
+    cdef double violation = 0
+    cdef double r_actual_sign = 1.0
+    cdef Py_ssize_t sample_idx
+
+    cdef Py_ssize_t r_col_size = 0
+    cdef Py_ssize_t r_actual_col_size = 0
+    cdef Py_ssize_t r_col_idx_idx = 0
+    cdef Py_ssize_t r_actual_col_idx_idx
+
+    cdef Py_ssize_t [:] r_col_indices = np.zeros(n_samples, dtype=np.intp)
+    cdef Py_ssize_t [:] r_actual_col_indices = np.zeros(n_samples, dtype=np.intp)
+
+    for sample_idx in prange(n_samples, nogil=True):
+        r_col_size = r_indptr[sample_idx + 1] - r_indptr[sample_idx]
+        r_actual_col_size = r_actual_indptr[sample_idx + 1] - r_actual_indptr[sample_idx]
+
+        r_col_indices[sample_idx] = 0
+        r_actual_col_indices[sample_idx] = 0
+
+        while r_col_indices[sample_idx] < r_col_size or r_actual_col_indices[sample_idx] < r_actual_col_size:
+            r_col_idx_idx = r_indices[
+                r_indptr[sample_idx]
+                + r_col_indices[sample_idx]
+            ]
+            r_actual_col_idx_idx = r_actual_indices[
+                r_actual_indptr[sample_idx]
+                + r_actual_col_indices[sample_idx]
+            ]
+
+            if r_col_idx_idx >= r_actual_col_idx_idx:
+                r_actual_sign = copysign(
+                    r_actual_sign,
+                    r_actual_data[
+                        r_actual_indptr[sample_idx]
+                        + r_actual_col_indices[sample_idx]
+                    ]
+                )
+
+                r_actual_data[
+                    r_actual_indptr[sample_idx]
+                    + r_actual_col_indices[sample_idx]
+                ] = fabs(
+                    r_actual_data[
+                        r_actual_indptr[sample_idx]
+                        + r_actual_col_indices[sample_idx]
+                    ]
+                ) - lambda_
+
+                r_actual_data[
+                    r_actual_indptr[sample_idx]
+                    + r_actual_col_indices[sample_idx]
+                ] = fmax(
+                    r_actual_data[
+                        r_actual_indptr[sample_idx]
+                        + r_actual_col_indices[sample_idx]
+                    ],
+                    0
+                )
+
+                if (
+                        r_actual_data[
+                            r_actual_indptr[sample_idx]
+                            + r_actual_col_indices[sample_idx]
+                        ] != 0
```

Comment:
but this is `double`, should be `fabs(...) < eps` instead

Reply:
Nope. The value is clipped so that it's either exactly equals 0 or not.